oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,89 @@
+    - Draft
+  * - Type
+    - Architecture
+  * - Created
+    - 2019-04-25
+
+Abstract
+========
+
+Proposes a key openedx component that will connect users to applications using a layer 7 (application layer/URL Based) routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each microservice occupies a unique hostname, with dedicated ELB or CDN infrastructure to support routing user traffic to the service. This pattern is simple to monitor, deploy, and manage, but slow and expensive to build, and regularly causes problems with cross service user information sharing. In Addition, modifying or splitting a monolithic service into microservices results in confusing URL changes and costly migrations.
+
+Specification
+=============
+
+A layer 7 routing application will be chosen that meets the following requirements:
+
+* High performance connection management capabilities, including support for slow clients or backends, and wire-speed throughput.
+* SSL termination on the router.
+* Flexible configuration backends including container service discovery, flat files, and filesystem routing.
+* Path based routing, including path subsets and regular expression matching
+
+This layer 7 routing application will be deployed on a common hostname.
+
+Rationale
+=========
+
+A layer 7 routing layer would allow us to separate the user facing URL or API endpoint from the backend implementation. It would allow microservices to exist that only manage a small number of endpoints, without requring developers to build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for the edx.org marketing site, allowing them to replace drupal generated pages with gatsby generated pages quickly and seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy new openedx sites quicker and easier than they currently can.  The router will provide a map for people to understand which services they need to deploy.
+
+
+
+Implementation
+==============
+
+* to be determined
+
+Future
+======
+
+Possible additional functionality that may be implemented at the l7 router:
+
+* Authentication
+* Rate limiting
+* Web Application Firewall
+
+
+Backwards Compatibility
+=======================
+
+Services will begin to move API and user endpoints behind this layer 7 router after declaring their top level routing path. eg.
+
+* the LMS would occupy /courses
+* prospectus would occupy /course, /school, /micromasters
+* ecommerce would occupy /orders, /payments
+* studio would occupy /studio 
+
+links between services would begin to point to the new endpoints, and old endpoints serve 301 redirects.
+
+Old URLs will be migrated using 301 redirects, which will automatically migrate users browsers to the new endpoints. For an overlapping time period, both URLs will be maintained and function, so that services can transition to the new schema.
```

Comment:
Are we talking a big-bang migration of everything, or will the new routing layer slowly see services ported over to it?

Reply:
I'm expecting a gradual migration.  First candidates will be new services.